### PR TITLE
[CL Incentives]: Sync global uptime accumulators upon tick crossing

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #4278

## What is the purpose of the change

This PR implements uptime-related logic for tick crossing. Specifically, it "flips" the tick-level uptime accum values upon crossing.

## Brief Changelog
 
- Implement logic for flipping uptime trackers when a tick is crossed

## Testing and Verifying

- New logic is relatively well tested in `tick_test.go` (including cases for technically undefined behavior such as crossing into non-adjacent ticks)

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/concentrated-liquidity/tick_test.go
Comment (stackman27) [inline x/concentrated-liquidity/tick_test.go:441]: Does "flip it" mean one side of the tick is cross so the liquidity is flipped to one token?

File: x/concentrated-liquidity/tick_test.go
Comment (stackman27) [inline x/concentrated-liquidity/tick_test.go:447]: shouldn't this name be "get tickInfo on existing pool and existing tick above current tick(nil uptime tracker)" 

i'm going off of this logic 
```
// uptimeGrowthOutside =
// { uptimeGrowthGlobal current tick >= tick }
// { 0                  current tick <  tick }
//
```

File: x/concentrated-liquidity/tick_test.go
Comment (stackman27) [inline x/concentrated-liquidity/tick_test.go:457]: same thing as above comment, if existingTick > currentTick, shouldn't we automatically expect nil uptime growth outside the ticks?

File: x/concentrated-liquidity/tick.go
Comment (stackman27) [inline x/concentrated-liquidity/tick.go:88]: does `updatedUptimeTrackers[uptimeId].UptimeGrowthOutside` hold uptime below lowerTick and above upperTick? I'm just trying to fit `secondsPerLiquidityInside = secondsPerLiquidity - secondsPerLiquidityBelow(lowerTick) - secondsPerLiquidityAbove(upperTick).` from the paper into what we're using here

**Review by stackman27 — COMMENTED** [review]
left few comments

File: x/concentrated-liquidity/tick.go
Comment (stackman27) [reply x/concentrated-liquidity/tick.go:88]: oooh are we just calculating one side of the equation (bolded part) , as described here

For example, if tickCurrent < i, then **secondsPerLiquidityBelow(i) = secondsPerLiquidity - secondsPerLiquidityOutside(i)**; if tickCurrent >= i, then secondsPerLiquidityBelow(i) = secondsPerLiquidityOutside(i).

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go:441]: It means switching a tick's uptime growth "outside" from `growth above the tick` to `growth below the tick` (or vice versa)

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go:447]: Ah good point, naming should be other way around

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go]: Fixed

File: x/concentrated-liquidity/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick.go:88]: Yep! SecondsPerLiquidityInside functionality is implemented in this PR: https://github.com/osmosis-labs/osmosis/pull/4407

File: x/concentrated-liquidity/incentives_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives_test.go:106]: This is to ensure that tests don't break if/when we decide to change the supported uptimes (specifically the number of supported uptimes). Everything is regenerated at runtime to stay compatible.

Once we finalize the specific uptimes we're supporting, we can replace this function with hard-coded values.

File: x/concentrated-liquidity/tick.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick.go:80]: Why do we need to update them as we cross them?

Comment (p0mvn) [reply x/concentrated-liquidity/tick.go:80]: Shouldn't we be able to update them once per block?

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick.go:80]: My thinking was that the update logic for the target tick's uptime accumulators will involve the global accum values, so we need to ensure they are up to date. Are you suggesting this isn't necessary?

I do see a different issue though – we should be updating the accums prior to getting them instead of after, will fix.

Comment (p0mvn) [reply x/concentrated-liquidity/tick.go:75]: So over a single swap, we might be crossing multiple ticks. From my understanding, we do not need to update the uptime accumulators every tick because that is idempotent within a swap. As a result, calling this update function as we cross every tick might be redundant.

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:455]: I know this existed before but I think we can merge these test cases into a single one because if a tick is found, we just return it without updates. The positions of the current tick relative to the active tick is irrelevant, and having two test cases only makes it more confusing

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:457]: Here, the tick did not exist so the name is misleading

Comment (p0mvn) [reply x/concentrated-liquidity/tick_test.go:457]: I also think that we can remove "on existing pool" part in the names of most test cases and just have one simple test case with the name "non existent pool - error".

Having it adds a cognitive overhead for every test case but this is just a  simple top-level check that is unimportant relative to other test case details

File: x/concentrated-liquidity/tick.go
Comment (p0mvn) [reply x/concentrated-liquidity/tick.go:75]: Outside of the scope of this PR but - why do we also need to call `updateUptimeAccumulatorsToNow` in `getTickInfo`?

It feels that this update needs to happen at another level of abstraction or in another location

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:431]: nit: let's remove on "existing pool" part from every test case name, please. It adds cognitive overhead while being unimportant context relative to other test case details because pool existence check happens once at the top level

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:457]: Isn't it ("zero initial uptime") trackers. Should the test case name reflect that?

File: x/concentrated-liquidity/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/tick_test.go:447]: Does it make sense also having a test case where non-zero initial uptime trackers and above current tick?

**Review by p0mvn — APPROVED** [review]
LGTM pending minor suggestions and the discussion on calling `updateUptimeAccumulatorsToNow` in `crossTick` and `getTickInfo`.

If that makes it easier for making progress, please feel free to merge this and move the discussion on Slack

File: x/concentrated-liquidity/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick.go:75]: > So over a single swap, we might be crossing multiple ticks. From my understanding, we do not need to update the uptime accumulators every tick because that is idempotent within a swap. As a result, calling this update function as we cross every tick might be redundant.

Good point – we can add a check so it only updates if `LastUpdateTime < current time`. Made an issue here (#4422) to not block this PR (and because it could break some downstream PRs)

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick.go:75]: > Outside of the scope of this PR but - why do we also need to call `updateUptimeAccumulatorsToNow` in `getTickInfo`?
> 
> It feels that this update needs to happen at another level of abstraction or in another location

We update there simply because `getInitialUptimeGrowthOutsidesForTick` uses global accum values (so they need to be up to date). The level of abstraction does seem a little off, but the function does include accumulator init logic more broadly so it might be worth discussing a small refactor or rename

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go:447]: This is already implicitly covered by case right below your comment, but I just added an explicit case for clarity anyway.

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go:431]: Agreed, done

File: x/concentrated-liquidity/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/tick_test.go:457]: Hmm not sure I follow – the uptime trackers are nonzero in this test case no?

**Pull request merged**
